bonito/openvino/model.py

```diff
@@ -0,0 +1,93 @@
+import os
+import numpy as np
+import torch
+
+try:
+    from openvino.inference_engine import IECore, StatusCode
+    from .loader import torch2openvino
+except ImportError:
+    pass
+
+class OpenVINOModel:
+
+    def __init__(self, model, half, dirname):
+        self.model = model
+        self.alphabet = model.alphabet
+
+        model_name = model.config['model'] + ('_fp16' if half else '')
+        xml_path, bin_path = [os.path.join(dirname, model_name) + ext for ext in ['.xml', '.bin']]
+        self.ie = IECore()
+        if os.path.exists(xml_path) and os.path.exists(bin_path):
+            self.net = self.ie.read_network(xml_path, bin_path)
+        else:
+            self.net = torch2openvino(model)
+        self.exec_net = None
+
+
+    def eval(self):
+        pass
+
+
+    def half(self):
+        return self
+
+
+    def to(self, device):
+        self.device = str(device).upper()
+
+
+    def __call__(self, data):
+        data = np.expand_dims(data, axis=2)  # 1D->2D
+        batch_size = data.shape[0]
```

Comment:
Is batch size placement fix for this framework?

Reply:
Currently the M.R supports `bonito evaluate` which takes fixed `N, C, W` chunks however some consideration is needed to support `bonito basecaller` where N=1 and W is variable for every batch.

Reply:
We process every sample in batch asynchronously to achieve the best efficiency.

Reply:
@iiSeymour, does `chunksize` affect basecaller accuracy? In example,
```
bonito basecaller dna_r9.4.1 /data/reads > basecalls.fasta --chunksize=3600
```
comparing to
```
bonito basecaller dna_r9.4.1 /data/reads > basecalls.fasta
```
?

Reply:
With a decent overlap the difference is negligible. 

```bash
bonito basecaller dna_r9.4.1 --chunksize=3600 --overlap 900  /data/reads > basecalls.fasta 
```

Reply:
We can try to enable it automatically (i.e. if OpenVINO model receives long read we split it y chunks with overlap and process effectively). Or add a note that better performance is achieved with chunks. What do you think?

BTW, why not to compute overlap automatically depends on chunksize? Because it depends on the number of paddings in model.

Reply:
I'm just working on batching chunks across multiple reads for maximum performance in the basecaller. 

yes, I think it makes sense to compute the overlap automatically based on the chunksize.

Reply:
In the following code I just tried to run read of size `[1, 1, 16154]` by chunks of size `[1, 1, 3600]`. Then stitch and compare. Unfortunately, there are multiple issues:
1. Valid area calculation requires to sum all the paddings (better sum paddings divided by strides) and know output size of the model
2. Last chunk problem - to achieve correct results we cannot shift it so it's not always matched chunk size (last chunk is excluded from the example)

<details>
<summary>example.py</summary>

```python
import torch
import numpy as np
from bonito.util import load_model

model = load_model('dna_r9.4.1', 'cpu', weights=0)
model.eval()

pads = 542
chunksize = 3600
out_size = 1200

inp = np.load('data.npy').reshape(1, 1, -1)
print('input shape:', inp.shape)

ref = model(torch.Tensor(inp))

pred = np.zeros_like(ref.detach().numpy())
length = out_size - 2 * pads
num = 2 + (inp.shape[-1] - chunksize) // (3 * length)
print('num chunks:', num)
chunks = []
for i in range(0, num-1):
    idx = i * length
    if i == 0:
        chunk = torch.Tensor(inp[:,:,:chunksize])
        out = model(chunk).detach().numpy()
        pred[0, :pads+length] = out[0, :pads+length]
    else:
        chunk = torch.Tensor(inp[:,:,3*idx:3*idx+chunksize])
        out = model(chunk).detach().numpy()
        pred[0, pads+idx:pads+idx+length] = out[0, pads:pads+length]

total = pads+idx+length
print('predicted:', total, '/', ref.shape[-1])
print(np.max(np.abs(ref.detach().numpy()[:,:total] - pred[:,:total])))
```

</details>